text/0000-standard-lazy-types.md

```diff
@@ -0,0 +1,461 @@
+[drawbacks]: #drawbacks
+
+* This is a moderately large addition to stdlib, there's a chance we do something wrong.
+  This can be mitigated by piece-wise stabilization (in particular, `Lazy` convenience types are optional) and the fact that API is tested in the crates.io ecosystem via `once_cell` crate.
+
+* The design of `Lazy` type uses default type-parameter as a work-around for the absence of type-inference of statics.
+
+* We use the same name for unsync and sync types, which might be confusing.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Why not `Lazy` as a primitive?
+
+On the first look, it may seem like we don't need `OnceCell`, and should only provide `Lazy`.
+The critical drawback of `Lazy` is that it's not always possible to provide the closure at creation time.
+
+This is important for lazy fields:
+
+```rust
+struct Ctx {
+    config_path: PathBuf,
+    config: Lazy<String, ???>,
+}
+
+impl Ctx {
+    pub fn new(config_path: PathBuf) -> Ctx {
+        Ctx {
+            config_path,
+            config: Lazy::new(|| {
+                // We would like to write something like
+                // `fs::read_to_string(&self.config_path)`
+                // here, but we can't have access to `self`
+                ???
+            })
+        }
+    }
+}
+```
+
+Or for singletons, initialized with parameters:
+
+```rust
+use std::{env, io, sync::OnceCell};
+
+#[derive(Debug)]
+pub struct Logger { ... }
+
+static INSTANCE: OnceCell<Logger> = OnceCell::new();
+impl Logger {
+    pub fn global() -> &'static Logger {
+        INSTANCE.get().expect("logger is not initialized")
+    }
+    fn from_cli(args: env::Args) -> Result<Logger, std::io::Error> { ... }
+}
+
+fn main() {
+    let logger = Logger::from_cli(env::args()).unwrap();
+
+    // Note how we use locally-created value for initialization.
+    INSTANCE.set(logger).unwrap();
+
+    // use `Logger::global()` from now on
+}
+```
+
+## Why `OnceCell` as a primitive?
+
+It is possible to imagine a type, slightly more general than `OnceCell`:
+
+```rust
+struct OnceFlipCell<U, V> { ... }
+
+impl<U, V> OnceFlipCell<U, V> {
+    const fn new(initial_value: U) -> OnceFlipCell<U, V>;
+
+    fn get_or_init<F: FnOnce(U) -> V>(&self, f: F) -> &V;
+}
+
+type OnceCell<T> = OnceFlipCell<(), T>;
+```
+
+That is, we can store some initial state in the cell and consume it during initialization.
+In practice, such flexibility seems to be rarely required.
+Even if we add a type, similar to `OnceFlipCell`, having a dedicated `OnceCell` (which *could* be implemented on top of `OnceFlipCell`) type simplifies a common use-case.
+
+## Variations of `set`
+
+The RFC proposes "obvious" signature for the `set` method:
+
+```rust
+fn set(&self, value: T) -> Result<(), T>;
+```
+
+Note, however, that `set` establishes an invariant that the cell is initialized, so a more precise signature would be
+
+```rust
+fn set(&self, value: T) -> (&T, Option<T>);
+```
```

Comment:
Wouldn't returning either a reference or the original value in the `Result` make more sense?
```rust
fn set(&self, value: T) -> Result<&T, T>;
```

I like this signature more, because it states the guarantee, and can allow users to avoid a `get` right after `set`.

It does not really cost us anything either. `set` already has to check an atomic to see if the cell is initialized, and creating the reference can be optimized away if it is not used.

Reply:
If the `set` fails, you still can get a reference back. That is, getting back back (old) &T and (new) T is a possible valid outcome of a blocking set operation

Reply:
O of course.
Some example code if you want to `set`, and then `get` a value.

The variant in the main text of the RFC:
```rust
pub fn set(&self, value: T) -> Result<(), T>;

let _unused = my_cell.set(42);
let val = my_cell.get();
```

Alternative variant in the RFC:
```rust
fn set(&self, value: T) -> (&T, Option<T>);

let (val, _unused) = my_cell.set(42);
```

My alternative:
```rust
fn set(&self, value: T) -> Result<&T, T>;

let val = match my_cell.set(42) {
    Ok(val) => val,
    Err(_unused) => my_cell.get(),
}
```

Okay, I prefer the basic variant in the RFC :+1: